src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/aarch64/macos/StackVaList.java

```diff
@@ -0,0 +1,282 @@
+        return new StackVaList(segment, List.of(segment), null);
+    }
+
+    static Builder builder(SharedUtils.Allocator allocator) {
+        return new Builder(allocator);
+    }
+
+    @Override
+    public void close() {
+        if (livenessCheck != null)
+            livenessCheck.close();
+        attachedSegments.forEach(MemorySegment::close);
+    }
+
+    @Override
+    public VaList copy() {
+        MemorySegment liveness = handoffIfNeeded(MemoryAddress.NULL.asSegmentRestricted(1),
+                segment.ownerThread());
+        return new StackVaList(segment, List.of(), liveness);
+    }
+
+    @Override
+    public VaList copy(NativeScope scope) {
+        Objects.requireNonNull(scope);
+        MemorySegment liveness = handoffIfNeeded(MemoryAddress.NULL.asSegmentRestricted(1),
+                segment.ownerThread());
+        liveness = liveness.handoff(scope);
+        return new StackVaList(segment, List.of(), liveness);
+    }
+
+    @Override
+    public MemoryAddress address() {
+        return segment.address();
+    }
+
+    @Override
+    public boolean isAlive() {
+        if (livenessCheck != null)
+            return livenessCheck.isAlive();
+        return segment.isAlive();
+    }
+
+    static class Builder implements VaList.Builder {
+
+        private final SharedUtils.Allocator allocator;
+        private final List<SimpleVaArg> args = new ArrayList<>();
+
+        public Builder(SharedUtils.Allocator allocator) {
+            this.allocator = allocator;
+        }
+
+        private Builder arg(Class<?> carrier, MemoryLayout layout, Object value) {
+            Objects.requireNonNull(layout);
+            Objects.requireNonNull(value);
+            SharedUtils.checkCompatibleType(carrier, layout, MacOsAArch64Linker.ADDRESS_SIZE);
+            args.add(new SimpleVaArg(carrier, layout, value));
+            return this;
+        }
+
+        @Override
+        public Builder vargFromInt(ValueLayout layout, int value) {
+            return arg(int.class, layout, value);
+        }
+
+        @Override
+        public Builder vargFromLong(ValueLayout layout, long value) {
+            return arg(long.class, layout, value);
+        }
+
+        @Override
+        public Builder vargFromDouble(ValueLayout layout, double value) {
+            return arg(double.class, layout, value);
+        }
+
+        @Override
+        public Builder vargFromAddress(ValueLayout layout, Addressable value) {
+            return arg(MemoryAddress.class, layout, value.address());
+        }
+
+        @Override
+        public Builder vargFromSegment(GroupLayout layout, MemorySegment value) {
+            return arg(MemorySegment.class, layout, value);
+        }
+
+        public VaList build() {
+            if (args.isEmpty()) {
+                return EMPTY;
+            }
+
+            // Each argument may occupy up to four slots
+            MemorySegment segment = allocator.allocate(VA_SLOT_SIZE_BYTES * args.size() * 4);
```

Comment:
This seems to confirm that always offsetting by VA_SLOT_SIZE_BYTES in the read and write code can be problematic, if arguments can occupy up to 4 slots, always offsetting the read/write cursor by one slot seems incorrect.